src/gui/qgsfieldmappingmodel.h

```diff
@@ -0,0 +1,165 @@
+     * and \a destinationFields, initial values for the expressions can be
+     * optionally specified through \a expressions which is a map from the original
+     * field name to the corresponding expression. A \a parent object
+     * can be also specified.
+     */
+    QgsFieldMappingModel( const QgsFields &sourceFields = QgsFields(),
+                          const QgsFields &destinationFields = QgsFields(),
+                          const QMap<QString, QgsExpression> &expressions = QMap<QString, QgsExpression>(),
+                          QObject *parent = nullptr );
+
+    //! Returns the context generator with the source fields
+    QgsExpressionContextGenerator *contextGenerator() const;
+
+    //! Returns TRUE if the destination fields are editable
+    bool destinationEditable() const;
+
+    //! Sets the destination fields editable state to \a editable
+    void setDestinationEditable( bool editable );
+
+    //! Returns a static map of supported data types
+    const QMap<QVariant::Type, QString> dataTypes() const;
+
+    //! Returns a list of source fields
+    QgsFields sourceFields() const;
+
+    //! Returns a list of Field objects representing the current status of the model
+    QList<QgsFieldMappingModel::Field> mapping() const;
+
+    //! Appends a new \a field to the model, with an optional \a expression
+    void appendField( const QgsField &field, const QgsExpression &expression = QgsExpression() );
+
+    //! Removes the field at \a index from the model, returns TRUE on success
+    bool removeField( const QModelIndex &index );
+
+    //! Moves down the field at \a index
+    bool moveUp( const QModelIndex &index );
+
+    //! Moves up the field at \a index
+    bool moveDown( const QModelIndex &index );
+
+    //! Set source fields to \a sourceFields
+    void setSourceFields( const QgsFields &sourceFields );
+
+    /**
+     * Set destination fields to \a destinationFields, initial values for the expressions can be
+     * optionally specified through \a expressions which is a map from the original
+     * field name to the corresponding expression.
+     */
+    void setDestinationFields( const QgsFields &destinationFields,
+                               const QMap<QString, QgsExpression> &expressions = QMap<QString, QgsExpression>() );
+
+    // QAbstractItemModel interface
+    int rowCount( const QModelIndex &parent = QModelIndex() ) const override;
+    int columnCount( const QModelIndex &parent = QModelIndex() ) const override;
+    QVariant data( const QModelIndex &index, int role ) const override;
+    QVariant headerData( int section, Qt::Orientation orientation, int role ) const override;
+    Qt::ItemFlags flags( const QModelIndex &index ) const override;
+    bool setData( const QModelIndex &index, const QVariant &value, int role ) override;
+
+  private:
+
+    class ExpressionContextGenerator: public QgsExpressionContextGenerator
```

Comment:
I don't think this is needed -- you could just make QgsFieldMappingModel itself a QgsExpressionContextGenerator and avoid this class.